### DISCUSSION TITLE
coreutils

### DISCUSSION THREAD

**Discussion by gngrossi** [zopencommunity/meta#265] [General]
Any additional utilities being considered?  df and file would be good candidates.
The man page for df appears like an updated df is available...I may be missing the port that contains it?
thanks

**Comment by IgorTodorovskiIBM** [comment]
Ah, `df` is part of coreutils. However, since it's not fully functional we haven't exposed it yet. @MikeFultonDev we should consider filtering the man pages too for coreutils.

FYI, the current issue with df is this:
```
./df
df: cannot read table of mounted file systems: EDC5129I No such file or directory.
```

**Comment by IgorTodorovskiIBM** [comment]
We have a few fixes for who, users, uptime, sha*sum, md5, and od coming up. But this is a good reminder to do another pass and see which tools are functional enough to be exposed.

**Reply by v1gnesh** [reply]
Please also consider `tail`. The `tail` that come with zOS doesn't work for datasets.
That is, `tail "//'HLQ.FILE1'"` doesn't work. But strangely, `head` works for datasets.

**Comment by gngrossi** [comment]
I see groups and shred are available.
thanks

**Comment by gngrossi** [comment]
Has there been a recent update to the core utilities?  Are the commands that "replace" the ones in /bin fully compatible?
thanks

**Reply by IgorTodorovskiIBM** [reply]
Hey Gary, no updates since last time. I believe `cp` is the only command that has been extended to support datasets. The rest are essentially vanilla ports. We are currently prioritizing the next iteration of `zopen`, but if you feel this is important from your perspective, then please let us know.

**Comment by gngrossi** [comment]
Hello Igor,
I have thought the coreutils are a foundation for USS. Since these are likely sourced ahead of /bin, they become very important for compatibility. What is the difference between a vanilla port and the /bin version? Is the plan to keep that or they will be extended for the way you see the USS utilities in the future (64 bit ascii enabled)?  thanks

cat
cp
date
echo
fmt
groups
install
join
md5sum
mktemp
numfmt
od
pinky
ptx
readlink
realpath
sha1sum
sha224sum
sha256sum
sha384sum
sha512sum
shred
sort
touch
tr
vdir
wc
yes

**Comment by MikeFultonDev** [comment]
I opened up an issue for this: https://github.com/ZOSOpenTools/coreutilsport/issues/48

**Comment by MikeFultonDev** [comment]
If someone wants to grab this, it is likely a pretty simple change.

**Comment by MikeFultonDev** [comment]
'cp' has been partially addressed @gngrossi by checking for '//' and if present 'failing over' to '/bin/cp'. There is an issue to do the same for 'cat'. The intent is for the tools to be a proper 'superset' of the UNIX System Services tool, but for some like 'cp' that is a relatively daunting task (hence the fail over). 
If you know of particular tools that cause grief, please comment.

**Comment by gngrossi** [comment]
Is this expected?
bash-5.2$ cp -O c=utf-8 "//'@02858.T.COZ.SAMPLIB(QQQ)'" GG11
cp: FSUMF192 fcntl() failed and could not set autoconversion for GG11: EDC5121I Invalid argument.

bash-5.2$ /bin/cp -O c=utf-8 "//'@02858.T.COZ.SAMPLIB(QQQ)'" GG11
cp: FSUMF192 fcntl() failed and could not set autoconversion for GG11: EDC5121I Invalid argument.

# works
bash-5.2$ cp -O c=utf-8 "//'@02858.T.S3TEST.SPECIAL'" GG10
bash-5.2$ /bin/cp -O c=utf-8 "//'@02858.T.S3TEST.SPECIAL'" GG10a

**Comment by MikeFultonDev** [comment]
@gngrossi looks like a bug - can you open one against coreutils and assign it to me? i'll take a peek...

**Comment by MikeFultonDev** [comment]
coreutils build has been fixed so that it only keeps man pages for tools we install.

**Comment by gngrossi** [comment]
Are there additional utilities being planned for coreutils?

**Comment by MikeFultonDev** [comment]
I would say 'yes' - it's voted high from the community that we add new coreutils.
If you have specific ones, I would recommend either a discussion entry or an issue to voice your opinion.
